File: src/main/kotlin/at/searles/parsing/tools/properties/PojoCreator.java
package at.searles.parsing.tools.properties;

import at.searles.parsing.Mapping;
import at.searles.parsing.ParserStream;
import org.jetbrains.annotations.NotNull;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

/**
 * Creates an object with setters and getters but without knowing
 * the parameters. No inversion thus.
 */
public class PojoCreator<T> implements Mapping<Properties, T> {

    private Constructor<T> ctor;
    private Class<T> clazz;

    public PojoCreator(Class<T> clazz, boolean withInfo) {
        try {
            this.ctor = withInfo ? clazz.getConstructor(ParserStream.class) : clazz.getConstructor();
            this.clazz = clazz;
        } catch (NoSuchMethodException e) {
            throw new IllegalArgumentException(e);
        }
    }

    public PojoCreator(Class<T> clazz) {
        this(clazz, false);
    }

    @Override
    public T parse(ParserStream stream, @NotNull Properties left) {
        try {
            T obj = ctor.getParameterCount() == 0 ? ctor.newInstance() : ctor.newInstance(stream);

            for(String key: left) {
                Method setter = MethodUtils.setter(clazz, key);
                setter.invoke(obj, (Object) left.get(key));
            }

            return obj;
        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
            throw new IllegalArgumentException(e);
        }
    }
}

File: src/main/kotlin/at/searles/parsingtools/properties/PojoCreator.kt
package at.searles.parsingtools.properties

import at.searles.parsing.Mapping
import at.searles.parsing.ParserStream

import java.lang.reflect.Constructor
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method

/**
 * Creates an object with setters and getters but without knowing
 * the parameters. No inversion thus.
 */
class PojoCreator<T> @JvmOverloads constructor(clazz: Class<T>, withInfo: Boolean = false) : Mapping<Properties, T> {

    private var ctor: Constructor<T>? = null
    private var clazz: Class<T>? = null

    init {
        try {
            this.ctor = if (withInfo) clazz.getConstructor(ParserStream::class.java) else clazz.getConstructor()
            this.clazz = clazz
        } catch (e: NoSuchMethodException) {
            throw IllegalArgumentException(e)
        }

    }

    override fun parse(stream: ParserStream, left: Properties): T? {
        try {
            val obj = if (ctor!!.parameterCount == 0) ctor!!.newInstance() else ctor!!.newInstance(stream)

            for (key in left) {
                val setter = MethodUtils.setter(clazz!!, key)
                setter.invoke(obj, left[key])
            }

            return obj
        } catch (e: InstantiationException) {
            throw IllegalArgumentException(e)
        } catch (e: IllegalAccessException) {
            throw IllegalArgumentException(e)
        } catch (e: InvocationTargetException) {
            throw IllegalArgumentException(e)
        } catch (e: NoSuchMethodException) {
            throw IllegalArgumentException(e)
        }

    }
}
